File: programs/list/client/src/instruction/createList.ts
import { Program } from "@project-serum/anchor";
import {
  PublicKey,
  SystemProgram,
  TransactionInstruction,
} from "@solana/web3.js";
import { ListProgram } from "../idl";
import { Account } from "../account";

export type CreateListArgs = {
  namespace: PublicKey;
  owner: PublicKey;
  payer: PublicKey;
};

export class CreateList {
  private account: Account;
  private program: Program<ListProgram>;

  constructor(account: Account, program: Program<ListProgram>) {
    this.account = account;
    this.program = program;
  }

  public async createList({
    namespace,
    owner,
    payer,
  }: CreateListArgs): Promise<TransactionInstruction> {
    const listPDA = await this.account.list.pda(owner, namespace);
    return this.program.instruction.createList(listPDA.bump, {
      accounts: {
        list: listPDA.address,
        namespace: namespace,
        owner: owner,
        payer: payer,
        systemProgram: SystemProgram.programId,
      },
    });
  }
}

File: programs/list/src/instructions/create_list.rs
use {
    anchor_lang::{
        prelude::*,
        solana_program::system_program
    },
    crate::state::*,
    std::mem::size_of
};

#[derive(Accounts)]
#[instruction(
    bump: u8,
)]
pub struct CreateList<'info> {
    #[account(
        init, 
        seeds = [
            SEED_LIST, 
            owner.key().as_ref(),
            namespace.key().as_ref()
        ],
        bump = bump, 
        payer = payer, 
        space = 8 + size_of::<List>()
    )]
    pub list: Account<'info, List>,

    #[account()]
    pub namespace: AccountInfo<'info>,

    #[account(mut)]
    pub owner: Signer<'info>,

    #[account(mut)]
    pub payer: Signer<'info>,
    
    #[account(address = system_program::ID)]
    pub system_program: Program<'info, System>,
}

pub fn handler(
    ctx: Context<CreateList>, 
    bump: u8,
) -> ProgramResult {
    // Get accounts.
    let list = &mut ctx.accounts.list;
    let namespace = &ctx.accounts.namespace;
    let owner = &ctx.accounts.owner;

    // Initialize list account.
    list.owner = owner.key();
    list.namespace = namespace.key();
    list.count = 0;
    list.bump = bump;
    
    return Ok(());
}
